gbasis/spherical.py

```diff
@@ -0,0 +1,286 @@
+    if not isinstance(mag, int):
+        raise TypeError("The magnetic quantum number must be an integer.")
+    if np.abs(mag) > angmom:
+        raise ValueError("The magnetic quantum number must be between -(angmom) and angmom.")
+
+    harmonic = {}
+    norm = harmonic_norm(angmom, mag)
+
+    for i, j, k in [
+        (x, y, z + shift_factor(mag))
+        for x in range((angmom - np.abs(mag)) // 2 + 1)
+        for y in range(x + 1)
+        for z in range(np.abs(mag) // 2 + 1)
+    ]:
+        # Multiply by n to easily make transform. matrix
+        coeff = expansion_coeff(angmom, mag, i, j, k) * norm
+        # a_x, a_y, a_z are the components of a Cartesian Gaussian primitive
+        a_x = int(2 * i + np.abs(mag) - (2 * (j + k)))
+        a_y = int(2 * (j + k))
+        a_z = int(angmom - 2 * i - np.abs(mag))
+        if coeff != 0:
+            harmonic[(a_x, a_y, a_z)] = coeff
+
+    return harmonic
+
+
+def generate_transformation(angmom, cartesian_order):
+    """Generate the transformation matrix for a given shell.
+
+    The rows of the matrix correspond to the Cartesian primitives, ordered as given
+    in cartesian_order. The columns of the matrix correspond to the spherical Gaussians,
+    ordered in ascending order from -(angmom) to angmom.
+
+    Parameters
+    ----------
+    angmom : int
+        The angular momentum of the basis function(s).
+
+    cartesian_order : list(len = 2 * angmom + 1) of tuple(len = 3)
+        The order of the Cartesian primitives, with each tuple holding
+        components in the order a_x, a_y, a_z.
+
+    Returns
+    -------
+    transform : np.ndarray((angmom + 1)*(angmom + 2) / 2, 2 * angmom + 1)
+        The transformation matrix from Cartesian primitives to spherical primitives.
+
+    Raises
+    ------
+    TypeError
+        If angmom is not an integer.
+        If cartesian_order is not a list
+        If each member of cartesian_order is not a tuple.
+    ValueError
+        If angmom is negative.
+        If cartesian_order does not have (angmom + 1)*(angmom + 2) / 2 components.
+        If any tuple in cartesian_order does not have 3 components.
+        If the components of any primitive do not sum to angmom.
+
+    """
+    if not isinstance(angmom, int):
+        raise TypeError("Angular momentum must be an integer.")
+    if angmom < 0:
+        raise ValueError("Angular momentum must be a non-negative integer.")
+    if not isinstance(cartesian_order, list):
+        raise TypeError("The order of the Cartesian primitives must be a list.")
+    if not all(isinstance(t, tuple) for t in cartesian_order):
+        raise TypeError("The Cartesian primitives' components must be a tuple.")
+    test_order = np.array(cartesian_order)
+    try:
+        if test_order.shape[1] != 3:
+            raise ValueError("Each Cartesian primitive must have three components.")
+    except IndexError:
+        raise ValueError("Each Cartesian primitive must have exactly three components.")
```

Comment:
I think it would be better to check the `cartesian_order` as a list rather than converting it to a numpy array. There might be some weird behaviour if someone decides to give a list of tuple of other iterables.

I would do the following:
- `if not all (all(isinstance(i, int) for i in t) for t in cartesian_order)` to check that all the elements inside the tuple is an integer
- `if not all (len(t) == 3 for t in cartesian_order` to check number of components
- then you can change to a numpy array